pkg/vm/engine/tae/containers/vector.go

```diff
@@ -345,73 +370,344 @@ func (vec *vector[T]) ReadFromFile(f common.IVFile, buffer *bytes.Buffer) (err e
+	// Hence, we are using src.Get(i) to retrieve the Null value as such from the src, and inserting
+	// it into the current CnVectorAdapter via this function.
+	for i := srcOff; i < srcOff+srcLen; i++ {
+		vec.Append(src.Get(i))
+	}
+}
+
+func (vec *vector[T]) CloneWindow(offset, length int, allocator ...*mpool.MPool) Vector {
+	opts := Options{}
+	if len(allocator) == 0 {
+		opts.Allocator = vec.GetAllocator()
+	} else {
+		opts.Allocator = allocator[0]
+	}
+
+	/*
+		// Approach 1: Deep Copy downstreamVector
+		// Problem: Data Race TestCRUD()
+		cloned := NewVector[T](vec.GetType(), vec.Nullable(), opts)
+		clonedTaeVector, _ := cnVector.Dup(vec.downstreamVector, vec.GetAllocator())
+		defer clonedTaeVector.Free(vec.GetAllocator())
+
+		clonedNsp := vec.downstreamVector.Nsp.Clone()
+
+		for i := offset; i < offset+length; i++ {
+			isNull := clonedNsp.Contains(uint64(i))
+
+			if isNull {
+				cloned.Append(types.Null{})
+			} else {
+				val := GetNonNullValue(clonedTaeVector, uint32(i))
+				cloned.Append(val)
+			}
+		}
+		return cloned
+
+		// Approach 2: ForEach Window
+		// Problem: Data Race TestCRUD()
+		cloned := NewVector[T](vec.GetType(), vec.Nullable(), opts)
+		op := func(v any, _ int) error {
+			cloned.Append(v)
+			return nil
+		}
+		err := vec.ForeachWindow(offset, length, op, nil)
+		if err != nil {
+			return nil
+		}
+		return cloned
+
+		// Approach 3: Deep copy DN vector
+		// Problem: Data Race TestCRUD()
+		window := NewVector[T](vec.GetType(), vec.Nullable(), opts)
+		window.ResetWithData(vec.Bytes(), vec.NullMask())
+		window.tryPromoting()
+		defer window.Close()
+
+		cloned := NewVector[T](window.GetType(), window.Nullable(), opts)
+		op := func(v any, _ int) error {
+			cloned.Append(v)
+			return nil
+		}
+		err := window.ForeachWindow(offset, length, op, nil)
+		if err != nil {
+			return nil
+		}
+
+		// Approach 4: CN Vector Window()
+		// Problem: panic: internal error: mp header corruption in TestTxn6()
+
+		clone, _ := cnVector.Dup(vec.downstreamVector, vec.GetAllocator())
+
+		window := cnVector.New(vec.GetType())
+		window = cnVector.Window(clone, offset, offset+length, window)
+
+		result := NewVector[T](vec.GetType(), vec.Nullable(), opts)
+		result.downstreamVector = window
+		result.isOwner = true
+
+		return result
+
+		// Approach 5: CN Dup + CN Shrink
+		// Problem: Data Race TestCRUD()
+		clonedTaeVector, _ := cnVector.Dup(vec.downstreamVector, vec.GetAllocator())
+		var sels []int64
+		for i := int64(offset); i < int64(offset)+int64(length); i++ {
+			sels = append(sels, i)
+		}
+		cnVector.Shrink(clonedTaeVector, sels)
+		cloned.downstreamVector = clonedTaeVector
+		cloned.isOwner = true
+
+		Main Problem: data-race issue in
+		1. cnVector.Dup()
+		2. ForEach Allocate( Get(i) )
+	*/
+
+	cloned := NewVector[T](vec.GetType(), vec.Nullable(), opts)
+	if vec.GetType().IsVarlen() {
+		//TODO: Data Race problem for VarLen
+
+		clonedMoVec, _ := cnVector.Dup(vec.downstreamVector, vec.GetAllocator())
```

Comment:
I am unable to solve the data-race issue for `CloneWindow()` on `Varlen Vector`. I have listed all the failed approaches I have tried. Any suggestions would be appreciated!